File: hw06.py
from math import cos, sin, sqrt, atan, atan2
import numpy as np

def linktrans(a, d, alpha, theta):
    ctheta = cos(theta)
    stheta = sin(theta)
    calpha = cos(alpha)
    salpha = sin(alpha)
    return np.array(
      [[ctheta, -stheta*calpha,   stheta*salpha, a*ctheta],
       [stheta,  ctheta*calpha,  -ctheta*salpha, a*stheta],
       [0,       salpha,          calpha,        d],
       [0,       0,               0,             1]] )

def fkine(ai, di, alpha, theta):
    mat = linktrans(ai[0], di[0], alpha[0], theta[0])
    for i in range(1,len(ai)):
        mat = np.matmul(mat, linktrans(ai[i], di[i], alpha[i], theta[i]))
    return mat


def revmat(theta, alpha):
    ctheta = cos(theta)
    stheta = sin(theta)
    calpha = cos(alpha)
    salpha = sin(alpha)
    return np.array(
      [[ctheta, -stheta*calpha,  stheta*salpha],
       [stheta,  ctheta*calpha, -ctheta*salpha],
       [     0,         salpha,         calpha]]
    )

def ikinebaxter(a1,a2,d1,d4,d6,T,LR,UD,NF,alpha=[-np.pi/2, 0, -np.pi/2, np.pi/2, -np.pi/2, 0]):
    pi = np.pi
    theta = np.zeros(6)
    R06 = T[:-1,:-1]
    d06 = T[:-1,-1]

# %   find d04
    d56 = np.array([0, 0, d6])
    d56 = np.matmul(R06, d56)
    d04 = d06 - d56

# %   find theta 1-3

    x = d04[0]
    y = d04[1]
# %   LR decides lefty or righty
    if LR == 1:
        theta[0] = atan2(y, x)
    else:
        theta[0] = atan(y / x)

    R01 = revmat(theta[0], alpha[0])

    d01 = np.array([a1*cos(theta[0]), a1*sin(theta[0]), d1])
    d14 = np.matmul(R01.T, (d04 - d01))
    x = d14[0]
    y = d14[1]

    xy = x**2 + y**2
    ad = a2**2 + d4**2
    ad2 = 2*a2*d4
    temp = (ad2 - ad + xy) / (ad2 + ad - xy)
    theta[2] = 2 * atan( sqrt( temp ) )
# %   whether you choose + or minus needs to depend on UD (elbow up or down)
    if UD == 1:
        theta[2] = -theta[2]
    theta[2] = pi - theta[2]

    phi = atan2(y, x)
    omega = atan2( (d4*sin(theta[2])), (a2 + d4*cos(theta[2])) )
    theta[1] = phi - omega

    theta[2] = (theta[2] - pi/2)

# %   find theta 4-6

    R12 = revmat(theta[1], alpha[1])
    R23 = revmat(theta[2], alpha[2])
    R03 = np.matmul(np.matmul(R01, R12), R23)
    R36 = np.matmul(R03.T, R06)

    y = R36[1,2]
    x = R36[0,2]
    if (y == 0) and (x == 0):
        if True:
            theta[3] = atan2(R36[1,0], R36[0,0])
        else:
            theta[3] = atan2(-R36[1,0], -R36[0,0])

    else:
        theta[3] = atan2(-y, -x)
        if NF == 1:
            theta[3] = -theta[3]


        y = -R36[0,2]*cos(theta[3]) - R36[1,2]*sin(theta[3])
        x = R36[2,2]
        theta[4] = atan2(y, x)

        y = -R36[0,0]*sin(theta[3]) + R36[1,0]*cos(theta[3])
        x = -R36[1,0]*sin(theta[3]) + R36[1,1]*cos(theta[3])
        theta[5] = atan2(y, x)

    return theta

def get_d06(a1,a2,d1,d4,thetas):
    xy = (
      a1 + a2*cos(thetas[1]) + d4*(
        cos(thetas[2])*sin(thetas[1]) - sin(thetas[2])*cos(thetas[1])
      )
    )

    dx = cos(thetas[0])*( xy )

    dy = sin(thetas[0])*( xy )

    dz = (
      d1 + a2*sin(thetas[1]) + d4*(
        -cos(thetas[1])*cos(thetas[2]) - sin(thetas[1])
      )
    )

    return np.round([dx, dy, dz], 2)

if __name__ == '__main__':
    pi = np.pi
    acc = 0

    ais = [0, 1, 0, 0, 0, 0]
    dis = [1, 0, 0, 1, 0, 0]
    alphas = [-pi/2, 0, -pi/2, pi/2, -pi/2, 0]

    print('test 1')
    thetas = np.zeros(6)
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', [1, 0, 0])
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )


    print('\ntest 2')
    thetas = [pi/2, -pi/2, 0, 0, 0, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', [0, 1, 2])
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )

    print('\ntest 3')
    thetas = [pi/4, -pi/4, 0, 0, 0, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', [1, 1, 1])
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )

    print('\ntest 4')
    dis = [2, 0, 0, 1, 0, 1]
    thetas = [pi/4, -pi/4, 0, -pi/4, 0, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', np.round([1.5, 1.5, 2-(sqrt(2)/2)], 2))
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )

    print('\ntest 5')
    ais = [1, 1, 0, 0, 0, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', np.round([1.5+(sqrt(2)/2), 1.5+(sqrt(2)/2), 2-(sqrt(2)/2)], 2))
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )

    print('\ntest 6')
    thetas = [pi/4, -pi/4, 0, 0, pi/4, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    print('d06:', np.round([1+(sqrt(2)/2), 1+(sqrt(2)/2), 1], 2))
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )



    print('\ntest 7')
    ais = [69, 370.82, 0, 0, 0, 0]
    dis = [270.35, 0, 0, 374.29, 0, 0]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 0, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    acc += (thets == thetas).sum() / 6
    # print('d06c:', get_d06(ais[0], ais[1], dis[0], dis[3], thetas) )

    print('\nactual')
    ais = [69, 370.82, 0, 0, 0, 0]
    dis = [270.35, 0, 0, 374.29, 0, 368.2]
    alphas = [-pi/2, 0, -pi/2, pi/2, -pi/2, 0]
    thetas = [pi/2, -pi/4, pi/3, pi, pi/6, pi/2]
    mat = fkine(ais, dis, alphas, thetas)
    thets = ikinebaxter(ais[0], ais[1], dis[0], dis[3], dis[5], mat, 1, 0, 1, alphas)
    print('thetas:', np.round(thetas, 2))
    print('fkine:', np.round(mat[:-1,-1], 2))
    print('ikine:', np.round(thets, 2))
    acc += (thets == thetas).sum() / 6

    print('\naccuracy:', (acc / 8))

File: hw06.m
% linktrans
%
%   linktrans(a,d,alpha,theta)
%   takes in the dh parameters and returns a homogenous matrix 0T1 that
%   maps from 0 to 1.
%
%   a:float | distance along xi from zi-1 to zi
%   d:float | distance along zi-1 from xi-1 to xi
%   alpha:float | radian angle around xi from zi-1 to zi
%   theta:float | radian angle around zi-1 from xi-1 to xi
%
%   Iain Lee
%   Robotics
%   10-22-20


% test
% temp = linktrans(1, 1, 0, pi/2)
% temp2 = linktrans(0, 2, pi/2, 0)
% comb = mtimes(temp, temp2)
% comb2 = fkine([1 0], [1 2], [0 pi/2], [pi/2 0])
a = [69 370.82 0 0 0 0];
d = [270.35 0 0 374.29 0 368.2];
alpha = [-pi/2 0 -pi/2 pi/2 -pi/2 0];
theta = [pi/2 -pi/4 pi/3 pi pi/6 pi/2]

T = fkine(a, d, alpha, theta)

% calculated answer:
% T = [-1.0000   -0.0000   -0.0000   -0.0000;
%      -0.0000    0.9659    0.2588  329.6331;
%       0.0000    0.2588   -0.9659 -184.6309;
%            0         0         0    1.0000];

theta = ikinebaxter(a(1), a(2), d(1), d(4), d(6), T, 1, 0, 1)


function T = linktrans(a,d,alpha,theta)

    ctheta = cos(theta);
    stheta = sin(theta);

    calpha = cos(alpha);
    salpha = sin(alpha);

    T = [ctheta -stheta*calpha   stheta*salpha a*ctheta;
         stheta  ctheta*calpha  -ctheta*salpha a*stheta;
         0       salpha          calpha        d;
         0       0               0             1];

end


function T = fkine(a,d,alpha,theta)

    i = 1;
    oT = linktrans(a(i), d(i), alpha(i), theta(i));
    for i=2:length(a)
        nT = linktrans(a(i), d(i), alpha(i), theta(i));
        oT = mtimes(oT, nT);
    end
    T = oT;

end

function R = rotmatzx(theta, alpha)
    R = [cos(theta) -sin(theta)*cos(alpha)  sin(theta)*sin(alpha);
         sin(theta)  cos(theta)*cos(alpha) -cos(theta)*sin(alpha);
                  0             sin(alpha)             cos(alpha)];
end

function thetas = ikinebaxter(a1,a2,d1,d4,d6,T,LR,UD,NF)
    alphas = [-pi/2 0 -pi/2 pi/2 -pi/2 0];
    thetas = zeros(6,0);
    R0_6 = T(1:3,1:3);
    d0_6 = T(1:3,4);

%   check workspace
    maxdist = a2 + d4 + d6;
    xydist = sqrt(d0_6(1)^2 + d0_6(2)^2);
    if xydist > (maxdist+a1)
        "return None"
    end
    if d0_6(3) > (maxdist+d1)
        "return None"
    end

%   find d0_4
    z0_6 = R0_6(1:3,3);
    d0_4 = d0_6 - d6*z0_6;

%   find thetas 1-3
    x1_1 = [1 0 0].';
    z0_0 = [0 0 1].';
%   LR decides lefty or righty
    if LR == 1
      thetas(1) = atan2(d0_4(2), d0_4(1));
    else
      thetas(1) = atan(d0_4(2) / d0_4(1));
    end
    R0_1 = rotmatzx(thetas(1), alphas(1));
    d1_4 = mtimes( R0_1.', (d0_4 - d1*z0_0) ) - a1*x1_1;

    x = d1_4(1);
    y = d1_4(2);
    a1h = a2;
    a2h = d4;
    a12p = (a1h + a2h)^2;
    a12m = (a1h - a2h)^2;
    xy2 = x^2 + y^2;
    top = a12p - xy2;
    bot = xy2 - a12m;
    thetas(3) = 2 * atan( sqrt( top / bot ) );
%   whether you choose + or minus needs to depend on UD (elbow up or down)
    if UD == 1
      thetas(3) = -thetas(3);
    end
%     thetas(3) = pi - thetas(3);

    phi = atan2(y, x);
    yh = a2h * sin(thetas(3));
    xh = a1h + a2h*cos(thetas(3));
    psi = atan2(yh, xh);
    thetas(2) = phi - psi;

    thetas(3) = thetas(3) - pi/2;

%   find thetas 4-6
    R1_2 = rotmatzx(thetas(2), alphas(2));
    R2_3 = rotmatzx(thetas(3), alphas(3));
    R0_3 = mtimes(R0_1, mtimes(R1_2, R2_3));
    R3_6 = mtimes(R0_3.', R0_6);

    y = R3_6(2,3);
    x = R3_6(1,3);
    thetas(4) = atan(-R3_6(2,3) / -R3_6(1,3));
    if NF == 0
      if thetas(4) > (pi/2)
          thetas(4) = thetas(4) - pi;
      elseif thetas(4) < (-pi/2)
          thetas(4) = thetas(4) + pi;
      end
    else
      if thetas(4) < (pi/2) & thetas(4) > (-pi/2)
        thetas(4) = thetas(4) + pi;
      end
    end

    yh = -R3_6(1,3)*cos(thetas(4)) - R3_6(2,3)*sin(thetas(4));
    xh = R3_6(3,3);
    thetas(5) = atan2(yh, xh);

    yh = -R3_6(1,1)*sin(thetas(4)) + R3_6(2,1)*cos(thetas(4));
    xh = -R3_6(1,2)*sin(thetas(4)) + R3_6(2,2)*cos(thetas(4));
    thetas(6) = atan2(yh, xh);

end
